File: backend/app/auth_service.py
"""Firebase Auth helpers.

Sign-up uses the Admin SDK (so we control the uid and write the users/{uid}
profile doc atomically). Sign-in uses the Firebase Auth REST API to obtain an
ID token / refresh token to hand back to the client."""
import httpx
from fastapi import HTTPException
from firebase_admin import auth as admin_auth, firestore

from .config import FIREBASE_API_KEY
from .firebase import get_auth, get_db

_REST = "https://identitytoolkit.googleapis.com/v1/accounts"
_SECURE_TOKEN = "https://securetoken.googleapis.com/v1/token"

_FRIENDLY = {
    "EMAIL_NOT_FOUND": "No account found with that email",
    "INVALID_PASSWORD": "Incorrect password",
    "INVALID_LOGIN_CREDENTIALS": "Incorrect email or password",
    "USER_DISABLED": "This account has been disabled",
    "MISSING_PASSWORD": "Password is required",
}


async def _rest_sign_in(email: str, password: str) -> dict:
    if not FIREBASE_API_KEY:
        raise HTTPException(500, "FIREBASE_API_KEY is not configured on the server")
    async with httpx.AsyncClient(timeout=20) as client:
        resp = await client.post(
            f"{_REST}:signInWithPassword",
            params={"key": FIREBASE_API_KEY},
            json={"email": email, "password": password, "returnSecureToken": True},
        )
    if resp.status_code != 200:
        code = resp.json().get("error", {}).get("message", "AUTH_FAILED")
        raise HTTPException(401, _FRIENDLY.get(code, "Login failed"))
    return resp.json()


async def sign_up(name: str, email: str, password: str, phone: str | None):
    auth = get_auth()
    try:
        user = auth.create_user(email=email, password=password, display_name=name)
    except admin_auth.EmailAlreadyExistsError:
        raise HTTPException(409, "An account with this email already exists")
    except ValueError as e:
        raise HTTPException(400, f"Invalid signup details: {e}")
    except Exception as e:
        raise HTTPException(400, f"Could not create account: {e}")

    get_db().collection("users").document(user.uid).set({
        "email": email,
        "name": name,
        "phone": phone or "",
        "role": "patient",
        # Sign-up requires completing the email OTP (see auth router), so the
        # email is verified by the time we reach here.
        "emailVerified": True,
        "createdAt": firestore.SERVER_TIMESTAMP,
    })

    tokens = await _rest_sign_in(email, password)
    return user.uid, tokens


async def change_password(uid: str, email: str | None, current_password: str,
                          new_password: str) -> None:
    """Verify the current password, then set a new one for this account."""
    if not email:
        raise HTTPException(400, "This account has no email/password login")
    # Re-authenticate to confirm the current password (raises 401 if wrong).
    await _rest_sign_in(email, current_password)
    try:
        get_auth().update_user(uid, password=new_password)
    except Exception as e:
        raise HTTPException(400, f"Could not update password: {e}")


async def login(email: str, password: str) -> dict:
    return await _rest_sign_in(email, password)


async def refresh_id_token(refresh_token: str) -> dict:
    """Exchange a refresh token for a fresh ID token. Note: this endpoint
    returns snake_case keys (id_token, refresh_token, expires_in, user_id)."""
    if not FIREBASE_API_KEY:
        raise HTTPException(500, "FIREBASE_API_KEY is not configured on the server")
    async with httpx.AsyncClient(timeout=20) as client:
        resp = await client.post(
            _SECURE_TOKEN,
            params={"key": FIREBASE_API_KEY},
            data={"grant_type": "refresh_token", "refresh_token": refresh_token},
        )
    if resp.status_code != 200:
        raise HTTPException(401, "Could not refresh session; please sign in again")
    return resp.json()

File: frontend/lib/services/auth_service.dart
import 'package:firebase_auth/firebase_auth.dart';

import 'package:frontend/services/api_client.dart';
import 'package:frontend/services/token_store.dart';

class AuthResult {
  AuthResult({required this.uid, this.name, this.email});
  final String uid;
  final String? name;
  final String? email;
}

/// Handles sign-up / login / logout and token persistence.
class AuthService {
  AuthService(this._api, this._tokens);
  final ApiClient _api;
  final TokenStore _tokens;

  Future<AuthResult> signup({
    required String name,
    required String email,
    required String password,
    String? phone,
  }) async {
    final data = await _api.post('/auth/signup', body: {
      'name': name,
      'email': email,
      'password': password,
      if (phone != null && phone.isNotEmpty) 'phone': phone,
    });
    return _persist(Map<String, dynamic>.from(data as Map));
  }

  Future<AuthResult> login({
    required String email,
    required String password,
  }) async {
    final data = await _api.post('/auth/login', body: {
      'email': email,
      'password': password,
    });
    return _persist(Map<String, dynamic>.from(data as Map));
  }

  Future<AuthResult> _persist(Map<String, dynamic> data) async {
    final user = (data['user'] as Map?) ?? const {};
    await _tokens.saveTokens(
      idToken: data['idToken'] as String?,
      refreshToken: data['refreshToken'] as String?,
      uid: user['uid'] as String?,
    );
    return AuthResult(
      uid: (user['uid'] ?? '') as String,
      name: user['name'] as String?,
      email: user['email'] as String?,
    );
  }

  /// Send a registration OTP to [email]. Returns the raw response, which may
  /// include `devCode` when the backend has no SMTP configured (dev mode).
  Future<Map<String, dynamic>> sendOtp(String email) async {
    final data = await _api.post('/auth/send-otp', body: {'email': email});
    return Map<String, dynamic>.from(data as Map);
  }

  Future<bool> verifyOtp(String email, String code) async {
    final data = await _api.post(
      '/auth/verify-otp',
      body: {'email': email, 'code': code},
    );
    return data is Map && data['verified'] == true;
  }

  /// Change the signed-in user's password (verifies [current] server-side).
  Future<void> changePassword(String current, String newPassword) async {
    await _api.post('/auth/change-password', auth: true, body: {
      'currentPassword': current,
      'newPassword': newPassword,
    });
  }

  /// Flag the signed-in user's email as verified after they complete the OTP.
  Future<void> markEmailVerified() async {
    await _api.post('/auth/verify-email', auth: true);
  }

  Future<bool> get hasSession => _tokens.hasSession;

  Future<void> logout() async {
    try {
      await FirebaseAuth.instance.signOut();
    } catch (_) {/* Firebase not initialized / no session */}
    await _tokens.clear();
  }
}
